crates/pallet-grandpa-finality-verifier/src/lib.rs

```diff
@@ -157,17 +184,126 @@ pub mod pallet {
+            hash == finality_proof.block,
+            Error::<T>::InvalidJustification
+        );
+
+        // fetch current authority set
+        let authority_set = <CurrentAuthoritySet<T>>::get(chain_id);
+        let voter_set =
+            VoterSet::new(authority_set.authorities).ok_or(Error::<T>::InvalidAuthoritySet)?;
+        let set_id = authority_set.set_id;
+
+        // verify justification
+        verify_justification::<C::Header>((hash, *number), set_id, &voter_set, &justifications)
+            .map_err(|e| {
+                log::error!(
+                    target: "runtime::grandpa-finality-verifier",
+                    "Received invalid justification for {:?}: {:?}",
+                    hash,
+                    e,
+                );
+                Error::<T>::InvalidJustification
+            })?;
+
+        // Update any next authority set if any
+        let next_header = block.block.header;
+        try_enact_authority_change::<T, C>(chain_id, &next_header, set_id)?;
+
+        // Update best finalized header
+        BestFinalized::<T>::insert(chain_id, next_header.encode());
+        Ok(())
+    }
+
+    /// Check the given header for a GRANDPA scheduled authority set change. If a change
+    /// is found it will be enacted immediately.
+    ///
+    /// This function does not support forced changes, or scheduled changes with delays
+    /// since these types of changes are indicative of abnormal behavior from GRANDPA.
+    pub(crate) fn try_enact_authority_change<T: Config, C: Chain>(
+        chain_id: T::ChainId,
+        header: &C::Header,
+        current_set_id: sp_finality_grandpa::SetId,
+    ) -> DispatchResult {
+        // We don't support forced changes - at that point governance intervention is required.
+        ensure!(
+            find_forced_change(header).is_none(),
+            Error::<T>::UnsupportedScheduledChange
+        );
+
+        if let Some(change) = super::find_scheduled_change(header) {
+            // GRANDPA only includes a `delay` for forced changes, so this isn't valid.
+            ensure!(
+                change.delay == Zero::zero(),
+                Error::<T>::UnsupportedScheduledChange
+            );
+
+            let next_authorities = AuthoritySet {
+                authorities: change.next_authorities,
+                set_id: current_set_id + 1,
+            };
+
+            // Since our header schedules a change and we know the delay is 0, it must also enact
+            // the change.
+            CurrentAuthoritySet::<T>::insert(chain_id, &next_authorities);
+
+            log::info!(
+                target: "runtime::grandpa-finality-verifier",
+                "Transitioned from authority set {} to {}! New authorities are: {:?}",
+                current_set_id,
+                current_set_id + 1,
+                next_authorities,
+            );
+        };
+
+        Ok(())
+    }
+}
+
+impl<T: Config> FeedValidator<T::ChainId> for Pallet<T> {
```

Comment:
You mean move the trait defintion from feeds to runtime ?

Reply:
ahhh yeah. make sense

Reply:
This is will not make the build any better if the initialization of the chain comes as part of the feed since there are not calls to verifier pallet anymore. Some structures needs to be shared among them. so either we move those structs to common promitive crate or make these depend on each other

Reply:
What data structures need to be shared exactly? I thought it is all just vectors of bytes as far as pallet-feeds is concerned.

Reply:
Yeah it is, but i'm not happy with the vec or slices TBH